src/ralph/backends/data/s3.py

```diff
@@ -0,0 +1,386 @@
+        if target is None:
+            target = self.default_bucket_name
+
+        objects_to_skip = set()
+        if new:
+            objects_to_skip = set(self.get_command_history(self.name, "read"))
+
+        try:
+            paginator = self.client.get_paginator("list_objects_v2")
+            page_iterator = paginator.paginate(Bucket=target)
+            for objects in page_iterator:
+                if "Contents" not in objects:
+                    continue
+                for obj in objects["Contents"]:
+                    if new and f"{target}/{obj['Key']}" in objects_to_skip:
+                        continue
+                    if details:
+                        obj["LastModified"] = obj["LastModified"].isoformat()
+                        yield obj
+                    else:
+                        yield obj["Key"]
+        except ClientError as err:
+            error_msg = err.response["Error"]["Message"]
+            msg = "Failed to list the bucket %s: %s"
+            logger.error(msg, target, error_msg)
+            raise BackendException(msg % (target, error_msg)) from err
+
+    @enforce_query_checks
+    def read(
+        self,
+        *,
+        query: Union[str, BaseQuery] = None,
+        target: str = None,
+        chunk_size: Union[None, int] = None,
+        raw_output: bool = False,
+        ignore_errors: bool = False,
+    ) -> Iterator[Union[bytes, dict]]:
+        """Read an object matching the `query` in the `target` bucket and yields it.
+
+        Args:
+            query: (str or BaseQuery): The ID of the object to read.
+            target (str or None): The target bucket containing the objects.
+                If target is `None`, the `default_bucket` is used instead.
+            chunk_size (int or None): The chunk size for reading objects.
+            raw_output (bool): Controls whether to yield bytes or dictionaries.
+            ignore_errors (bool): If `True`, errors during the read operation
+                will be ignored and logged. If `False` (default), a `BackendException`
+                will be raised if an error occurs.
+
+        Yields:
+            dict: If `raw_output` is False.
+            bytes: If `raw_output` is True.
+
+        Raises:
+            BackendException: If a failure during the read operation occurs and
+                `ignore_errors` is set to `False`.
+            BackendParameterException: If a backend argument value is not valid and
+                `ignore_errors` is set to `False`.
+        """
+        if query.query_string is None:
+            msg = "Invalid query. The query should be a valid object name."
+            logger.error(msg)
+            raise BackendParameterException(msg)
+
+        if not chunk_size:
+            chunk_size = self.default_chunk_size
+
+        if target is None:
+            target = self.default_bucket_name
+
+        try:
+            response = self.client.get_object(Bucket=target, Key=query.query_string)
```

Comment:
I'm curious about the `get_object` method - is the returned `StreamingBody()` fetched in lazy or greedy manner?
If it's greedy, we might want to use (in future) the [download_fileobj](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/s3/client/download_fileobj.html) method, which can be configured with `multipart_chunksize`.

Reply:
It seems to return a `StreamingBody` stream that you would then `iter_lines` or `iter_chunks` over.
Wouldn't using `download_fileobj` means storing the whole file into memory? It would be faster though indeed, for files that can fit into memory.

Reply:
You are right, indeed the interface of the `download_fileobj` unfortunately doesn't let us yield results by chunks - `get_object` should be the right choice